src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRefreshQueue.cs

```diff
@@ -141,21 +149,52 @@ public async Task TryEnqueueRefreshComputationAsync(Project project, Cancellatio
 
     private void OnLspSolutionChanged(object? sender, WorkspaceChangeEventArgs e)
     {
-        if (e.DocumentId is not null && e.Kind is WorkspaceChangeKind.DocumentChanged)
-        {
-            var document = e.NewSolution.GetRequiredDocument(e.DocumentId);
-            var documentUri = document.GetURI();
+        Uri? documentUri = null;
 
+        if (e.DocumentId is not null)
+        {
             // We enqueue the URI since there's a chance the client is already tracking the
             // document, in which case we don't need to send a refresh notification.
             // We perform the actual check when processing the batch to ensure we have the
             // most up-to-date list of tracked documents.
-            EnqueueSemanticTokenRefreshNotification(documentUri);
-        }
-        else
-        {
-            EnqueueSemanticTokenRefreshNotification(documentUri: null);
+            if (e.Kind is WorkspaceChangeKind.DocumentChanged)
+            {
+                var document = e.NewSolution.GetRequiredDocument(e.DocumentId);
+                documentUri = document.GetURI();
+            }
+            else if (e.Kind is WorkspaceChangeKind.AdditionalDocumentChanged)
+            {
+                var document = e.NewSolution.GetRequiredAdditionalDocument(e.DocumentId);
+
+                // Changes to files with certain extensions (eg: razor) shouldn't trigger semantic a token refresh
+                if (DisallowsAdditionalDocumentChangedRefreshes(document.FilePath))
+                    return;
+            }
+            else if (e.Kind is WorkspaceChangeKind.DocumentReloaded)
+            {
+                var newDocument = e.NewSolution.GetRequiredDocument(e.DocumentId);
+                var oldDocument = e.OldSolution.GetDocument(e.DocumentId);
+
+                // If the document's attributes haven't changed, then use the document's URI for
+                //   the call to EnqueueSemanticTokenRefreshNotification which will enable the
+                //   tracking check before sending the WorkspaceSemanticTokensRefreshName message.
+                if (oldDocument?.State.Attributes is IChecksummedObject oldChecksumObject
+                    && newDocument.State.Attributes is IChecksummedObject newChecksumObject
+                    && oldChecksumObject.Checksum == newChecksumObject.Checksum)
+                {
+                    documentUri = newDocument.GetURI();
+                }
+            }
         }
+
+        EnqueueSemanticTokenRefreshNotification(documentUri);
+    }
+
+    // Duplicated from Microsoft.CodeAnalysis.LanguageServer.HostWorkspace.LoadedProject.TreatAsIsDynamicFile
+    private static bool DisallowsAdditionalDocumentChangedRefreshes(string? filePath)
+    {
+        var extension = Path.GetExtension(filePath);
+        return extension is ".cshtml" or ".razor";
```

Comment:
some context for this - 
It seems like in general additional files changing could affect semantic tokens (for example if those changes are inputs to a source generator).  However, Razor semantic tokens only uses the design time file so we actually don't care of the Razor file changes (they'll already tell us when the design time file changes).